scenario/smartsearch.go

```diff
@@ -0,0 +1,350 @@
+}
+
+func (settings *SmartSearchSettings) UnmarshalJSON(bytes []byte) error {
+	err := json.Unmarshal(bytes, &settings.SmartSearchSettingsCore)
+	if err != nil {
+		return err
+	}
+	switch settings.SearchTextSource {
+	case SearchTextSourceList:
+	case SearchTextSourceFile:
+		rowFile, err := helpers.NewRowFile(settings.SearchTextFilePath)
+		if err != nil {
+			return err
+		}
+		settings.SearchTextList = rowFile.Rows()
+	default:
+		return errors.Errorf("Unknown SearchTextSource<%d>", settings.SearchTextSource)
+	}
+	return nil
+}
+
+// Validate implements ActionSettings interface
+func (settings SmartSearchSettings) Validate() ([]string, error) {
+	var warnings []string
+	if len(settings.SearchTextList) == 0 {
+		return warnings, errors.New("no searchtext found")
+	}
+	for idx, searchtext := range settings.SearchTextList {
+		searchTerms := parseSearchTerms(searchtext)
+		if len(searchTerms) == 0 {
+			return warnings, errors.Errorf(`no search terms found in searchtext%d<%s> `, idx+1, settings.SearchTextList[idx])
+		}
+	}
+
+	return warnings, nil
+}
+
+func intPtr(i int) *int {
+	return &i
+}
+
+func quote(s string) string {
+	return fmt.Sprintf(`%#v`, s)
+}
+
+func quoteList(strList []string) string {
+	if len(strList) == 0 {
+		return "NONE"
+	}
+	quoted := []string{}
+	for _, v := range strList {
+		quoted = append(quoted, quote(v))
+	}
+	return strings.Join(quoted, ",")
+}
+
+func ternary(cond bool, ifCondTrue string, ifCondFalse string) string {
+	if cond {
+		return ifCondTrue
+	}
+	return ifCondFalse
+
+}
+
+func logSearchSuggestionResult(logEntry *logger.LogEntry, id int, searchSuggestionResult *enigma.SearchSuggestionResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	suggestions := []string{}
+	for _, suggestion := range searchSuggestionResult.Suggestions {
+		suggestions = append(suggestions, suggestion.Value)
+	}
+	logEntry.LogDebugf("search%d suggestions: %s", id, quoteList(suggestions))
+}
+
+func logSearchResult(logEntry *logger.LogEntry, id int, searchResult *enigma.SearchResult) {
+	if !logEntry.ShouldLogDebug() {
+		return
+	}
+	hasResult := false
+	for sgIdx, sga := range searchResult.SearchGroupArray {
+		for _, item := range sga.Items {
+			searchTermsMatched := []string{}
+			for _, stIndex := range item.SearchTermsMatched {
+				searchTermsMatched = append(searchTermsMatched, searchResult.SearchTerms[stIndex])
+			}
+			matches := []string{}
+			for _, match := range item.ItemMatches {
+				matches = append(matches, match.Text)
+			}
+			hasResult = true
+			logEntry.LogDebugf(`search%d result group%d matching term%s %s in %s %s: %s`, id, sgIdx+1,
+				ternary(len(searchTermsMatched) != 1, "s", ""), quoteList(searchTermsMatched),
+				item.ItemType, quote(item.Identifier), quoteList(matches))
+		}
+	}
+	if !hasResult {
+		logEntry.LogDebugf("search%d result: NONE", id)
+	}
+}
+
+var spaceRegex = regexp.MustCompile(`\s+`)
+
+func parseSearchTerms(s string) []string {
```

Comment:
We should get rid of this now that the search terms list has been changed to an array. No need to support splitting a string, right?

Reply:
No, it is used to separate the search text into terms, right before doing the ws rpc:s.